spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/DefaultOutputMessageBuilder.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * By default ({@link OutputMode#payload}) the inference score is stored directly within the output Message payload.
+ *
+ * The {@link OutputMode#header} stores the inference score inside a Message header with name {@link #outputName}.
+ * In this case the output message payload copies the inbound message payload.
+ *
+ * The {@link OutputMode#tuple} uses a {@link Tuple} to store the inference score along with input message payload.
+ * If the payload is already a tuple that contains an ORIGINAL_INPUT_DATA entry then copy the content of the input tuple
+ * in the new tuple to be returned.
+ *
+ * @author Christian Tzolov
+ */
+public class DefaultOutputMessageBuilder implements OutputMessageBuilder {
+
+	public static final String ORIGINAL_INPUT_DATA = "original.input.data";
+	public static final String APPLICATION_X_SPRING_TUPLE = "application/x-spring-tuple";
+
+	private OutputMode outputMode;
+
+	private final String outputName;
+
+	public DefaultOutputMessageBuilder(TensorflowCommonProcessorProperties properties) {
+		this.outputMode = properties.getMode();
+		this.outputName = properties.getOutputName();
+	}
+
+	@Override
+	public MessageBuilder<?> createOutputMessageBuilder(Message<?> inputMessage, Object computedScore) {
+		switch (this.outputMode) {
+
+		case tuple:
+			TupleBuilder outTupleBuilder = TupleBuilder.tuple().put(this.outputName, computedScore);
+
+			Object inputPayload = inputMessage.getPayload();
+
+			if (inputPayload instanceof Tuple && ((Tuple) inputPayload).hasFieldName(ORIGINAL_INPUT_DATA)) {
+				// If the payload is already a tuple that contains ORIGINAL_INPUT_DATA entry then copy the
+				// content of the input tuple in the new tuple to be returned.
+				outTupleBuilder.putAll((Tuple) inputPayload);
+			}
+			else {
+				// This is a new tuple so preserve the input data.
+				outTupleBuilder.put(ORIGINAL_INPUT_DATA, inputPayload);
+			}
+
+			return MessageBuilder
+					.withPayload(outTupleBuilder.build())
+					.setHeader(MessageHeaders.CONTENT_TYPE, APPLICATION_X_SPRING_TUPLE);
+
+		case header:
+			MessageBuilder<?> builder = MessageBuilder
+					.withPayload(inputMessage.getPayload())
+					.setHeader(this.outputName, computedScore);
+			if (inputMessage.getHeaders().get(MessageHeaders.CONTENT_TYPE) != null) {
+				builder.setHeader(MessageHeaders.CONTENT_TYPE, inputMessage.getHeaders().get(MessageHeaders.CONTENT_TYPE));
+			}
+			return builder;
+
+		default: // payload mode
+			return MessageBuilder
+					.withPayload(computedScore)
+					.setHeader(MessageHeaders.CONTENT_TYPE, findContentType(computedScore));
+		}
+
+	}
+
+	private String findContentType(Object computedScore) {
+		if (computedScore instanceof Tuple) {
+			return APPLICATION_X_SPRING_TUPLE;
+		}
+		else if (computedScore instanceof String) {
+			return MimeTypeUtils.APPLICATION_JSON_VALUE;
+		}
+
+		// default
+		return MimeTypeUtils.APPLICATION_JSON_VALUE;
```

Comment:
Hm. This default is the same as a variant for the `String`.  
Maybe something is missed in the logic?

Reply:
+1